### DISCUSSION TITLE
migrated from cocos2dx 3.17

### DISCUSSION THREAD

**Discussion by NgVThangBz** [axmolengine/axmol#1347] [Q&A]
Before in cocos2dx 3.17. my proj have inherited the draw() function and do as below
![Screenshot 2023-09-21 at 16 56 35](https://github.com/axmolengine/axmol/assets/125545428/757521af-d788-4cd5-87db-1010b3b58214)
but now with axmol they cause errors, what can I do to replace them

**Comment by rh101** [comment]
It's perhaps best to actually include the errors in your post.

**Reply by NgVThangBz** [reply]
![Screenshot 2023-09-21 at 17 05 50](https://github.com/axmolengine/axmol/assets/125545428/afea30a5-950b-429a-b5cb-a6f0176b0691)
this is my error,

**Reply by halx99** [reply]
only callback and group command need use command pool, others not use command pool

**Reply by NgVThangBz** [reply]
@halx99  Can you give specific examples?

**Reply by halx99** [reply]
what's your draw command type?

**Reply by halx99** [reply]
seems your command just use callback?

**Reply by halx99** [reply]
you should use callback command directly:
```cpp
renderer->addCallbackCommand([]{}, 0);
```

**Reply by NgVThangBz** [reply]
@halx99  Previously we used cocos2d::DrawPrimitives, I replaced 1 part with drawNode but still not completely

**Reply by NgVThangBz** [reply]
@halx99  do i need to call parent::draw()?

**Comment by NgVThangBz** [comment]
I also had a few problems, so I'll cover them all here:
- I use `drawNode->drawLine()`
```
_drawNode->clear();
_drawNode->setLineWidth(width);
_drawNode->drawLine()
```
With any width value, lines are always drawn at thickness 1.
in axmol `setLineWidth` is replaced or not.

**Reply by halx99** [reply]
Because opengl3.3 core profile not support linewidth > 1, and metal not support line width, so for consistent behavior, axmol2 limit linwidth=1 for all platforms, if you want thinkness > 1 line, recommanded use custom quad shader, or use LayerColor, actually godot draw lines with quad, means per line segemnt have 4 vertics

**Comment by NgVThangBz** [comment]
I use `EventListenerTouchAllAtOnce` to listen for layer touch events,
but it seems `touches` always = 1, how do I know if I'm touching the screen with 2 fingers?
![Screenshot 2023-09-22 at 08 29 26](https://github.com/axmolengine/axmol/assets/125545428/9cf83b66-a3f2-404f-a652-f33b2de1d596)
![Screenshot 2023-09-22 at 08 40 03](https://github.com/axmolengine/axmol/assets/125545428/510526ed-8c31-4b0b-8766-716a73f04dc8)

**Reply by NgVThangBz** [reply]
@halx99 do you think it bug?
